File: CDSA/circularQ.c
#include <stdio.h>
#include<stdlib.h>
#define size 5

struct circularQueue{
 int items[size];
 int front;int rear;
};
void initilizes(struct circularQueue* q){
    q->front=-1;
    q->rear=-1;
}
int isEmpty(struct circularQueue* q){
 return q->front==-1&&q->rear==-1;
}
int isFull(struct circularQueue* q){
    return (q->rear+1)%size==q->front;
}
void enQueue(struct circularQueue* q,int data){
      if(isFull(q)){
        printf("queue is full");
        return;
      }
      if(isEmpty(q)){
        q->front=0;
      }
      q->rear=(q->rear+1)%size;
      q->items[q->rear]=data;
}
int deQueue(struct circularQueue* q){
    if(isEmpty(q)){
        printf("q is empty");
        return -1;
    }
    int result=q->items[q->front];
    if(q->front==q->rear){
        q->front=-1;
        q->rear=-1;
    }else{
     q->front=(q->front+1)%size;
    }
    return result;
}
int peek(struct circularQueue* q){
    return q->items[q->front];
}
// Function to display the elements of the circular queue
void displayQueue(struct circularQueue* q) {
    if (isEmpty(q)) {
        printf("Queue is empty!\n");
        return;
    }
    printf("Queue elements: ");
    int i = q->front;
    while (1) {
        printf("%d ", q->items[i]);
        if (i == q->rear) {
            break;
        }
        i = (i + 1) % size;
    }
    printf("\n");
}
int main(){
    struct circularQueue q;
    initilizes(&q);
    enQueue(&q,10);
     enQueue(&q,20);
      enQueue(&q,30);
      displayQueue(&q);
      while(!isEmpty(&q)){
        printf("%d\n",peek(&q));
        deQueue(&q);
      }
}

File: LINKEDLIST/circularq.java
public class circularq{

    public static class queue{
        static int arr[] ;
        static  int size;
        static  int front;
        static int rear;
        public queue(int n){
            arr=new int[n];
            size=n;
            front=-1;
            rear=-1;
        }

        public static boolean isEmpty(){
            return rear==-1 && front==-1;
        }
        public static boolean isFull(){
            return (rear+1)%size==front;
        }
        public static void enQueue(int data){
            if(isFull()){
                System.out.println("Q is full");
                return;
            }
            if(front==-1){
                front=0;
            }
            rear=(rear+1)%size;
            arr[rear]=data;
        
        }
        public static int dequeue(){
             if(isEmpty()){
                System.out.println("empty");
                return -1;
             }
             int val=arr[front];
             if(rear==front){
                rear=front=-1;
             }else{
                front=(front+1)%size;
             }
             return val;
        }
    }
    public static void main(String[] args) {
         queue q=new queue(5);
         q.enQueue(1);
         q.enQueue(2);
         q.enQueue(3);
         q.enQueue(4);
         q.enQueue(5);
         while(!q.isEmpty()){
         System.out.println(q.dequeue());
         }
    }
}